### ISSUE TITLE
ObjectPattern Containing ArrayPattern

### ISSUE THREAD

**Issue by jacobp100** [babel/babel#4741]
I’m trying to build the ast for the parameter of `({ key: [value] }) => {}`. I’m using the code below, which AST explorer seems to suggest is correct. However, running this throws (Property value of ObjectProperty expected node to be of a type [“Expression”] but instead got “ArrayPattern”).

Am I doing something wrong, or does the ObjectProperty type need extending?

``` js
// Attempting to define the parameter for
// ({ key: [value] }) => {};
t.ObjectPattern([
    t.ObjectProperty(
        t.Identifier('key'),
        t.ArrayPattern([t.Identifier('value')])
    )
]);
```

See https://runkit.com/jacobp100/types

**Comment by hzoo** [comment]
https://github.com/babel/babel/blob/478bba0adeabe51f0d0aaff3144d2b5d01c4798e/packages/babel-types/src/definitions/core.js#L529

Don't think it's wrong but we need to modify the validation check somehow (either changing ObjectProperty or ArrayPattern)

**Comment by dk00** [comment]
[`ArrayPattern` in `ArrayPattern`](//astexplorer.net/#/JLxdGtJGqh) also doesn't work.

```js
t.arrayPattern([
  t.identifier('a'),
  t.arrayPattern([t.identifier('b')])
])
```

```
TypeError: Property elements[1] of ArrayPattern expected node to be of
a type ["Expression"] but instead got "ArrayPattern"
```

**Comment by citycide** [comment]
I think this also covers what I'm running into, which is nested object patterns:

```js
t.objectPattern([
  t.objectProperty(
    t.identifier('a'),
    t.objectPattern([ /* ... */ ])
  )
])
```

I'm trying to build a destructured function parameter.

**Comment by hzoo** [comment]
Anyone want to make a PR to fix? (workaround is to use babel-template though)

**Comment by citycide** [comment]
@hzoo I can give it a shot and see if I can take care of it, will give a shout if I need to.

**Comment by citycide** [comment]
@hzoo Looks like a partial fix was merged in #5722. I can still patch up `ObjectProperty` but does that take care of `ArrayPattern` fully?
